stats.go

```diff
@@ -298,30 +301,95 @@ func (c *gauge) Value() uint64 {
 	return atomic.LoadUint64(&c.value)
 }
 
-type timer struct {
+type timer interface {
+	time(time.Duration)
+	AddDuration(time.Duration)
+	AddValue(float64)
+	AllocateSpan() Timespan
+	CollectedValue() []float64
+	SampleRate() float64
+}
+
+type standardTimer struct {
 	base time.Duration
 	name string
 	sink Sink
 }
 
-func (t *timer) time(dur time.Duration) {
+func (t *standardTimer) time(dur time.Duration) {
 	t.AddDuration(dur)
 }
 
-func (t *timer) AddDuration(dur time.Duration) {
+func (t *standardTimer) AddDuration(dur time.Duration) {
 	t.AddValue(float64(dur / t.base))
 }
 
-func (t *timer) AddValue(value float64) {
+func (t *standardTimer) AddValue(value float64) {
 	t.sink.FlushTimer(t.name, value)
 }
 
-func (t *timer) AllocateSpan() Timespan {
+func (t *standardTimer) AllocateSpan() Timespan {
 	return &timespan{timer: t, start: time.Now()}
 }
 
+func (t *standardTimer) CollectedValue() []float64 {
+	return nil // since we flush right away nothing will be collected
+}
+
+func (t *standardTimer) SampleRate() float64 {
+	return 1.0 // metrics which are not sampled have an implicit sample rate 1.0
+}
+
+type reservoirTimer struct {
+	base     time.Duration
+	name     string
+	capacity int
+	values   []float64
+	count    int
+	mu       sync.Mutex
```

Comment:
super nit, make the Mutex the first field (it's advantageous performance wise to make the most accessed field first)